### PR TITLE
2509 update focalloss to use sigmoid (7/July)

### LINKED ISSUE #2509: Focal Loss returning 0.0 with include_background=False (7/July)
**Describe the bug**
Focal Loss returns 0 for all entries when input consists of only one channel. Tried running the spleen 3D segmentation tutorial with `DiceFocalLoss(include_background=False)`. Printing out dice and focal loss separately shows the problem. 

I think the issue is that `logpt = F.log_softmax(i, dim=1)` returns 0 for arbritrary inputs. Should this be `F.logsigmoid()` in this case? 

**To Reproduce**
```
input = torch.randn(1, 1, 10)
F.log_softmax(input, dim=1)

tensor([[[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]]])
```

vs

```
input = torch.randn(1, 2, 10)
F.log_softmax(input, dim=1)

input = torch.randn(1, 2, 10)...
tensor([[[-2.6699, -0.4457, -1.1688, -0.3480, -0.9432, -0.8112, -0.2835,
          -2.2028, -1.3333, -1.8177],
         [-0.0718, -1.0226, -0.3722, -1.2245, -0.4933, -0.5876, -1.3990,
          -0.1171, -0.3060, -0.1772]]])
```

**Expected behavior**
Focal Loss should return correct log.probs. for single channel inputs. 

**Environment**

Printing MONAI config...

MONAI version: 0.6.dev2126
Numpy version: 1.19.5
Pytorch version: 1.9.0+cu102
MONAI flags: HAS_EXT = False, USE_COMPILED = False
MONAI rev id: 2ad54662de25e9a964c33327f7f2f178655573ef

Optional dependencies:
Pytorch Ignite version: NOT INSTALLED or UNKNOWN VERSION.
Nibabel version: 3.0.2
scikit-image version: 0.16.2
Pillow version: 7.1.2
Tensorboard version: 2.4.1
gdown version: 3.6.4
TorchVision version: 0.10.0+cu102
ITK version: NOT INSTALLED or UNKNOWN VERSION.
tqdm version: 4.41.1
lmdb version: 0.99
psutil version: 5.4.8
pandas version: 1.1.5

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

Fixes #2509

### Description
using sigmoid activation to align with the literature

the change is verified: 
previously gamma=0 reduces it to `nn.CrossEntropyLoss`
now gamma=0 reduces it to `nn.BCEWithLogitsLoss`

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Thanks for the quick fix.
@yiheng-wang-nv could you please help review it?

Thanks.

**Comment by yiheng-wang-nv** [comment]
Hi @wyli , in paper (https://arxiv.org/pdf/1708.02002.pdf) it uses `sigmoid` operation to compute `p` but it also says to use 2-class classification for instance:
Screen Shot 2021-07-05 at 10 58 15 AM
However, for multi-class task, I think it still needs to use `softmax` operation, just as the cross entropy loss does:
https://pytorch.org/docs/stable/_modules/torch/nn/modules/loss.html#CrossEntropyLoss

Hi @ristoh @ericspod , could you please help to double check it?

**Comment by wyli** [comment]
> https://arxiv.org/pdf/1708.02002.pdf

Hi @yiheng-wang-nv, the footnote you cited is for describing the formulation. In section 4 classification subnet, it's clear: "for each of the A anchors and K object classes... Finally sigmoid activations are attached to output the KA
binary predictions per spatial location"

replacing the sigmoid with softmax should work as well, for example in their source code they support both:
https://github.com/facebookresearch/Detectron/blob/1809dd41c1ffc881c0d6b1c16ea38d08894f8b6d/detectron/modeling/retinanet_heads.py#L281-L295

tensorflow implements the sigmoid version https://github.com/tensorflow/addons/blob/e83e71cf07f65773d0f3ba02b6de66ec3b190db7/tensorflow_addons/losses/focal_loss.py

**Comment by sandylaker** [comment]
> However, for multi-class task, I think it still needs to use `softmax` operation, just as the cross entropy loss does:
> https://pytorch.org/docs/stable/_modules/torch/nn/modules/loss.html#CrossEntropyLoss
> 
> Hi @ristoh @ericspod , could you please help to double check it?

I would like to add an extra comment. For anchor-based detector, sigmoid is more widely adopted. So the classification branch actually performs multi-label classification. At inference stage, each anchor box will be classified to the class with highest score, even when the scores do not fulfill `sum(softmax(scores)) = 1`. Hope it helps.

File: monai/losses/focal_loss.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
         Args:
             input: the shape should be BNH[WD], where N is the number of classes.
                 The input should be the original logits since it will be transferred by
```
Comment (ericspod) [inline monai/losses/focal_loss.py:80]: "transferred" -> "transformed"?

File: monai/losses/focal_loss.py [hunk 2]
```diff
@@ -142,11 +140,12 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
...
             at = at.expand((t.size(0), -1, t.size(2)))  # 1,N,1 => B,N,H*W
             # Multiply the log proba by their weights.
-            logpt = logpt * at
+            ce = ce * at
 
         # Compute the loss mini-batch.
-        weight = torch.pow(-pt + 1.0, self.gamma)
-        loss = torch.mean(-weight * t * logpt, dim=-1)
+        p = F.logsigmoid(-i * (t * 2.0 - 1.0))
+        loss = torch.mean((p * self.gamma).exp() * ce, dim=-1)
```
Comment (ericspod) [inline monai/losses/focal_loss.py:147]: Should this be `p**self.gamma`? Raising p to this power shows up in most other implementations of focal loss.

Comment (wyli) [reply monai/losses/focal_loss.py:147]: This version is adapted from a codebase used by our kaggle participants, it's reliable according to their recommendations... This `p` is a `logsigmoid` instead of `sigmoid`, `exp(log_p * gamma) = (exp(log_p))**gamma`, I think it'll reduce the change of overflow for `exp(x)` when x is large

Comment (ericspod) [reply monai/losses/focal_loss.py:147]: OK, this implementation isn't like anything else I'm finding while looking for focalloss implementations in Pytorch.

Comment (wyli) [reply monai/losses/focal_loss.py:147]: I had a quick search, a similar idea is presented here https://www.kaggle.com/rejpalcz/focalloss-for-keras

Comment (ericspod) [reply monai/losses/focal_loss.py:147]: Most exact implementation is [here](https://becominghuman.ai/investigating-focal-and-dice-loss-for-the-kaggle-2018-data-science-bowl-65fb9af4f36c) farther down the page.

File: monai/losses/focal_loss.py [hunk 3]
```diff
@@ -117,10 +117,8 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
         i = i.reshape(b, n, -1)
         t = t.reshape(b, n, -1)
 
-        # Compute the log proba.
-        logpt = F.log_softmax(i, dim=1)
-        # Get the proba
-        pt = torch.exp(logpt)  # B,H*W or B,N,H*W
+        max_val = (-i).clamp(min=0)
+        ce = i - i * t + max_val + ((-max_val).exp() + (-i - max_val).exp()).log()
```
Comment (ericspod) [inline monai/losses/focal_loss.py:121]: This is used in the other implementations but it's equivalent to [binary_cross_entropy_with_logits](https://github.com/pytorch/pytorch/blob/ac67cda272b0856c1be3ebfebc324e0b3bab286f/aten/src/ATen/native/Loss.cpp#L237) correct?

File: monai/losses/focal_loss.py (hunk 2 above)
Comment (wyli) [reply monai/losses/focal_loss.py:147]: that's almost the same as the one in this PR, correct? https://gist.github.com/AdrienLE/bf31dfe94569319f6e47b2de8df13416#file-focal_dice_1-py

File: monai/losses/focal_loss.py (hunk 3 above)
Comment (wyli) [reply monai/losses/focal_loss.py:121]: yeah looks like they are very similar...

File: monai/losses/focal_loss.py (hunk 2 above)
Comment (ericspod) [reply monai/losses/focal_loss.py:147]: It's the same gist actually. I think there's multiple sources that come back to the same place.

**Review by ericspod — APPROVED** [review]
I think it's good to go though perhaps some more explanation of the computation would help to clarify why this differs from the "standard" definitions of focal loss most commonly seen, why things are done in log space for instance.
